stl/inc/ranges

```diff
@@ -2635,6 +2589,421 @@ namespace ranges {
+                               && derived_from<typename _InnerTraits::iterator_category, forward_iterator_tag>,
+                        forward_iterator_tag,
+                        conditional_t<derived_from<typename _OuterTraits::iterator_category, input_iterator_tag>
+                                   && derived_from<typename _InnerTraits::iterator_category, input_iterator_tag>,
+                            input_iterator_tag,
+                            output_iterator_tag>>>;
+        };
+        // clang-format on
+
+        template <bool _Const>
+        class _Iterator : public _Category_base<iterator_traits<iterator_t<_Maybe_const<_Const, _Vw>>>,
+                              iterator_traits<iterator_t<range_reference_t<_Maybe_const<_Const, _Vw>>>>,
+                              is_reference_v<range_reference_t<_Maybe_const<_Const, _Vw>>>,
+                              common_range<range_reference_t<_Maybe_const<_Const, _Vw>>>> {
+        private:
+            template <bool>
+            friend class _Iterator;
+            template <bool>
+            friend class _Sentinel;
+
+            using _Parent_t  = _Maybe_const<_Const, join_view>;
+            using _Base      = _Maybe_const<_Const, _Vw>;
+            using _SubRng    = range_reference_t<_Base>;
+            using _OuterIter = iterator_t<_Base>;
+            using _InnerIter = iterator_t<_SubRng>;
+
+            // True if and only if the expression *i, where i is an iterator from the outer range, is a glvalue:
+            static constexpr bool _Deref_is_glvalue = is_reference_v<_SubRng>;
+
+            /* [[no_unique_address]] */ _OuterIter _Outer{};
+            /* [[no_unique_address]] */ _InnerIter _Inner{};
+            _Parent_t* _Parent{};
+
+            constexpr auto& _Update_inner(_SubRng _Range) {
+                if constexpr (_Deref_is_glvalue) {
+                    return _Range;
+                } else {
+                    _Parent->_Inner = views::all(_STD move(_Range));
+                    return _Parent->_Inner;
+                }
+            }
+
+            constexpr void _Satisfy() {
+                const auto _Last = _RANGES end(_Parent->_Range);
+                for (; _Outer != _Last; ++_Outer) {
+                    auto& _Tmp = _Update_inner(*_Outer);
+                    _Inner     = _RANGES begin(_Tmp);
+                    if (_Inner != _RANGES end(_Tmp)) {
+                        return;
+                    }
+                }
+                if constexpr (_Deref_is_glvalue) {
+                    _Inner = _InnerIter{};
+                }
+            }
+
+#if _ITERATOR_DEBUG_LEVEL != 0
+            constexpr void _Check_dereference() const noexcept {
+                _STL_VERIFY(_Parent != nullptr, "cannot dereference value-initialized join_view iterator");
+                _STL_VERIFY(_Outer != _RANGES end(_Parent->_Range), "cannot dereference end join_view iterator");

```

Comment:
I think we should write `join_view end iterator` rather than `end join_view iterator?`